source/client-side-operations-timeout/client-side-operations-timeout.rst

```diff
@@ -0,0 +1,875 @@
+
+If ``timeoutMS`` is not set and support for ``socketTimeoutMS`` has not been
+removed, drivers MUST honor ``socketTimeoutMS`` as the timeout for socket
+reads and writes.
+
+See `maxTimeMS accounts for server RTT`_.
+
+Batching
+~~~~~~~~
+
+If an operation must be sent to the server in multiple batches (e.g.
+``collection.bulkWrite()``), the ``timeoutMS`` option MUST apply to the
+entire operation, not to each individual batch.
+
+Retryability
+~~~~~~~~~~~~
+
+If an operation requires a retry per the retryable reads or writes
+specifications and ``timeoutMS`` is set to a non-zero value, drivers MUST
+retry operations as many times as possible before the timeout expires or a
+retry attempt returns a non-retryable error. Once the timeout expires, a
+timeout error MUST be raised.
+
+See `Why don’t drivers use backoff/jitter between retry attempts?`_.
+
+Client Side Encryption
+~~~~~~~~~~~~~~~~~~~~~~
+
+If automatic client-side encryption or decryption is enabled, the remaining
+``timeoutMS`` value MUST be used as the ``timeoutMS`` when executing
+``listCollections`` commands to retrieve collection schemas, ``find``
+commands to get data from the key vault, and any commands against
+mongocryptd. It MUST also be used as the request timeout for HTTP requests
+against KMS servers to decrypt data keys. When sending a command to
+mongocryptd, drivers MUST NOT append a ``maxTimeMS`` field. This is to ensure
+that a ``maxTimeMS`` field can be safely appended to the command after it has
+been marked by mongocryptd and encrypted by libmongocrypt. To determine
+whether or not the server is a mongocryptd, drivers MUST check that the
+``iscryptd`` field in the server's description is ``true``.
+
+For explicit encryption and decryption, the ``ClientEncryptionOpts`` options
+type used to construct `ClientEncryption
+<../client-side-encryption/client-side-encryption.rst#clientencryption>`_
+instances MUST support a new ``timeoutMS`` option, which specifies the timeout
+for all operations executed on the ``ClientEncryption`` object.
+
+See `maxTimeMS is not added for mongocryptd`_.
+
+Background Connection Pooling
+-----------------------------
+
+Connections created as part of a connection pool’s ``minPoolSize``
+maintenance routine MUST use ``connectTimeoutMS`` as the timeout for
+connection establishment. After the connection is established, if
+``timeoutMS`` is set at the MongoClient level, it MUST be used as the timeout
+for all commands sent as part of the MongoDB or authentication handshakes.
+The timeout MUST be refreshed after each command. These commands MUST set
+timeouts per the `Command Execution`_ section. If ``timeoutMS`` is not set,
+drivers MUST continue to honor ``socketTimeoutMS`` as the socket timeout for
+handshake and authentication commands.
+
+Server Monitoring
+-----------------
+
+Drivers MUST NOT use ``timeoutMS`` for commands executed by the server
+monitoring and RTT calculation threads.
```

Comment:
In the case of libmongoc, `isMaster` commands are sent out asynchronously and replies are read with a select/poll loop. The round itself will take as long as the slowest member. In practice, this can be `connectTimeoutMS` (default: 10 seconds) if a member is offline. For this reason, we advise users to tune their `connectTimeoutMS` value closer to their expected latency (e.g. 2x).

If libmongoc allows single-threaded monitoring to count towards `timeoutMS`, I can absolute see this causing problems. In an application where one node was perpetually offline but still part of the replica set configuration, I imagine a monitoring scan could regularly exhaust `timeoutMS` after each [`cooldownMS`](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-monitoring.rst#cooldownms) period has elapsed (and monitoring decides to try the server once again).

I agree there's no need to be prescriptive about how drivers implement the deadline. In the case of PHP wrapping libmongoc, communicating the timeout via an explicit session may be our only option.

Reply:
Do you have a suggestion for changing this? If doing `collection.insert(doc, timeoutMS=100)` is supposed to mean "insert this document in 100ms or give me back control if it takes longer than that" and libmongoc requires a foreground monitoring scan to block the operation's progress, I don't see anything we can do to alleviate that. Perhaps we should recommend that users increase `timeoutMS` when operating in single-threaded mode?

Reply:
> Perhaps we should recommend that users increase timeoutMS when operating in single-threaded mode?

I don't thinks that's tenable, but I understand if that's what we need to end up doing.

Users can't reasonably predict when SDAM monitoring might occur. I assume the user story of `timeoutMS` is saying, "I think this database operation shouldn't take longer than X seconds to complete and I'm willing to wait X, so I'll specify a deadline of X." That works for multi-threaded drivers with background monitoring, but for single-threaded we'd be asking them to always specify X + Y to cover a potential worst case (i.e. monitoring required and one or more servers are unreachable). By default, Y is 10 seconds in libmongoc. I think this would ultimately require any single-threaded user of `timeoutMS` to also fine-tune `connectTimeoutMS` and consider SDAM (which we've previously tried to insulate them from).

My proposal would be to permit single-threaded drivers to exclude SDAM from their `timeoutMS` calculation. However, I acknowledge that idea falls apart once we start thinking about how `timeoutMS` is really just wall clock time for things like `withTransaction` or GridFS stream ops. In that case, going out of the way to exclude SDAM makes less sense when we're still going to be counting time spent by _everything_ else (e.g. user's callback decides to call `sleep()`, GridFS reads from a slow user-provided stream).

Reply:
Can we capture the points in this thread somewhere later in the spec (e.g. Design Rationale or Q&A)? That way we have a record to note that we did consider how this might affect single-threaded SDAM and chose not to pursue a solution at this time. The justification is that `timeoutMS` is intentionally used to track wall time and that has the side effect of including potential single-threaded monitoring.